File: eel-core/src/main/scala/io/eels/component/parquet/ParquetIterator.scala
package io.eels.component.parquet

import com.typesafe.scalalogging.slf4j.StrictLogging
import io.eels.{InternalRow, Schema}
import io.eels.component.avro.{AvroRecordFn, AvroSchemaFn}
import org.apache.avro.generic.GenericRecord
import org.apache.parquet.hadoop.ParquetReader

object ParquetIterator extends StrictLogging {

  // creates a new parquet iterator. The schema must match the schema used in the reader
  def apply(reader: ParquetReader[GenericRecord], schema: Schema): Iterator[InternalRow] = new Iterator[InternalRow] {
    require(schema.columns.nonEmpty, "Attempted to create a parquet iterator with no schema columns")

    val recordFn = new AvroRecordFn
    val avroSchema = AvroSchemaFn.toAvro(schema)

    val iter = Iterator.continually(reader.read).takeWhile(_ != null).map { record =>
      val row = recordFn.fromRecord(record, avroSchema)
      require(row.size == schema.size, s"Row is missing values for some schema fields [row=$row; schema=$schema]")
      row
    }

    override def hasNext: Boolean = {
      val hasNext = iter.hasNext
      if (!hasNext) {
        reader.close()
      }
      hasNext
    }

    override def next(): InternalRow = iter.next()
  }
}

File: eel-core/src/main/kotlin/io/eels/component/parquet/ParquetIterator.kt
package io.eels.component.parquet

import io.eels.Row
import io.eels.component.avro.AvroRecordDeserializer
import io.eels.util.map
import io.eels.util.nullTerminatedIterator
import org.apache.avro.generic.GenericRecord
import org.apache.parquet.hadoop.ParquetReader
import java.util.stream.Stream

/**
 * Creates an Iterator<Row> that will return Row objects for each Parquet GenericRecord.
 * The Row objects returned will have the same schema as the parquet records.
 *
 * @reader the underlying parquet reader to use to load records
 *
 */
object ParquetIterator {
  operator fun invoke(reader: ParquetReader<GenericRecord>): Iterator<Row> = object : Iterator<Row> {

    val iter = Stream.generate { reader.read() }.nullTerminatedIterator().map { AvroRecordDeserializer().toRow(it) }

    override fun hasNext(): Boolean {
      val hasNext = iter.hasNext()
      if (!hasNext) {
        reader.close()
      }
      return hasNext
    }

    override fun next(): Row = iter.next()
  }
}